src/core_plugins/kibana/public/dashboard/grid.js

```diff
@@ -34,25 +68,25 @@ app.directive('dashboardGrid', function ($compile, Notifier) {
 
       // debounced layout function is safe to call as much as possible
       const safeLayout = _.debounce(layout, 200);
-
-      $scope.removePanelFromState = (panelIndex) => {
-        _.remove($scope.state.panels, function (panel) {
-          return panel.panelIndex === panelIndex;
-        });
-      };
-
       /**
-       * Removes the panel with the given index from the $scope.state.panels array. Does not
-       * remove the ui element from gridster - that is triggered by a watcher that is
-       * triggered on changes made to $scope.state.panels.
-       * @param panelIndex {number}
+       * Mapping of panelIndex to the angular element in the grid.
        */
-      $scope.getPanelByPanelIndex = (panelIndex) => {
-        return _.find($scope.state.panels, function (panel) {
-          return panel.panelIndex === panelIndex;
-        });
+      const panelElementMapping = {};
+
+      // Tell gridster to remove the panel, and cleanup our metadata
+      function removePanelFromGrid(panelElement, silent) {
+        // remove from grister 'silently' (don't reorganize after)
+        gridster.remove_widget(panelElement, silent);
+      }
+
+      $scope.removePanel = (panelIndex) => {
```

Comment:
Yep, exactly the issue I ran into, hence the slightly strange function names of removePanel and onPanelRemoved.  I actually started to go down the bindToController path but ended up backtracking because it became a bit too involved to add in to this PR, but will def investigate as a later step.

So my reasoning with the function names is that removePanel removes the panel from the UI, then it's saying to the parent, hey, a panel was removed, do what you need to do.  That is the reason for the present and then past tense voice.  If I kept them both present tense, removePanel and onPanelRemove, they are essentially synonymous.  My distinction is not a great one, and obviously leads to some confusion, but there was some thought behind it.

Given my explanation, if you still feel I should make both present tense, I will change it, though personally I think this is slightly better, and the bindToController option the best next step, and then giving them the same name, and same tense.